engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala

```diff
@@ -457,8 +533,31 @@ object PackageRepository {
     override def registerModuleCreatedInRuntime(module: Module): Unit =
       registerModule(module)
 
-    private def registerModule(module: Module): Unit =
+    private def registerModule(module: Module): Unit = {
       loadedModules.put(module.getName.toString, module)
+    }
+
+    /** Registering virtual module, unlike the non-compiler generated one, is conditional
+      * in a sense that if a module already exists with a given name we only update its
+      * list of virtual modules that it should export.
+      * If no module exists under the given name, we register the virtual one.
+      *
+      * @param virtualModule a virtual module to register
+      * @param refs list of names of modules that should be exported by the module under the given name
+      */
+    private def registerVirtualModule(
+      virtualModule: Module,
+      refs: List[QualifiedName]
+    ): Unit = {
+      if (!loadedModules.contains(virtualModule.getName.toString)) {
+        loadedModules.put(virtualModule.getName.toString, virtualModule)
+      } else {
+        val loaded = loadedModules(virtualModule.getName.toString)
+        if (!loaded.isVirtual && virtualModule.isVirtual) {
+          loaded.setDirectSubmoduleNames(refs.toArray)
```

Comment:
Maybe easier to convert `refs` to Java (unmodifiable) `List` rather than array.

Reply:
:+1: